components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/DefaultOIDCClaimsCallbackHandler.java

```diff
@@ -0,0 +1,628 @@
+        if (log.isDebugEnabled()) {
+            log.debug("Adding claims of user: " + authorizedUser + " to the JWTClaimSet used to build the id_token.");
+        }
+
+        try {
+            Map<String, Object> claims = getUserClaims(authzReqMessageContext);
+            setClaimsToJwtClaimSet(claims, jwtClaimsSet);
+        } catch (OAuthSystemException e) {
+            log.error("Error occurred while adding claims of user: " + authorizedUser + " to the JWTClaimSet used to " +
+                    "build the id_token.", e);
+        }
+    }
+
+    /**
+     * Get response map
+     *
+     * @param requestMsgCtx Token request message context
+     * @return Mapped claimed
+     * @throws OAuthSystemException
+     */
+    private Map<String, Object> getUserClaims(OAuthTokenReqMessageContext requestMsgCtx) throws OAuthSystemException {
+        // Get any user attributes that were cached against the access token
+        // Map<(http://wso2.org/claims/email, email), "peter@example.com">
+        Map<ClaimMapping, String> userAttributes = getUserAttributesCachedAgainstToken(getAccessToken(requestMsgCtx));
+        if (isEmpty(userAttributes)) {
+            userAttributes = getUserAttributesCachedAgainstAuthorizationCode(getAuthorizationCode(requestMsgCtx));
+            if (log.isDebugEnabled()) {
+                log.debug("Not claims cached against the access_token. Retrieving claims cached against the " +
+                        "authorization code.");
+            }
+        }
+        // Map<"email", "peter@example.com">
+        Map<String, Object> claims = getClaimMapForUser(requestMsgCtx, userAttributes);
+        String spTenantDomain = requestMsgCtx.getOauth2AccessTokenReqDTO().getTenantDomain();
+        // Restrict Claims going into the token based on the scope
+        return filterClaimsByScope(requestMsgCtx.getScope(), spTenantDomain, claims);
+    }
+
+    private String getAuthorizationCode(OAuthTokenReqMessageContext requestMsgCtx) {
+        return (String) requestMsgCtx.getProperty(AUTHZ_CODE);
+    }
+
+    private String getAccessToken(OAuthTokenReqMessageContext requestMsgCtx) {
+        return (String) requestMsgCtx.getProperty(ACCESS_TOKEN);
+    }
+
+    private Map<String, Object> getClaimMapForUser(OAuthTokenReqMessageContext requestMsgCtx,
+                                                   Map<ClaimMapping, String> userAttributes) {
+        Map<String, Object> claimMap = Collections.emptyMap();
+        if (isEmpty(userAttributes)) {
+            if (!isFederatedUser(requestMsgCtx)) {
+                if (log.isDebugEnabled()) {
+                    log.debug("No user attributes found in cache for user: " + requestMsgCtx.getAuthorizedUser() + "." +
+                            " Retrieving claims from userstore.");
+                }
+                return retrieveClaimsForLocalUser(requestMsgCtx, claimMap);
```

Comment:
do we need to pass claimMap as an argument here. Inside the method it assigns a new object